spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/repository/query/SpannerStatementQueryExecutor.java

```diff
@@ -169,32 +171,63 @@ private SpannerStatementQueryExecutor() {
 	 * @param <T> the type of the child persistent entity
 	 * @param writeConverter a converter to convert key values as needed to bind to the query
 	 *     statement.
+	 * @param mappingContext mapping context
 	 * @return the Spanner statement to perform the retrieval.
 	 */
 	public static <T> Statement getChildrenRowsQuery(Key parentKey,
-			SpannerPersistentEntity<T> childPersistentEntity, SpannerCustomConverter writeConverter) {
-		StringBuilder sb = new StringBuilder(
-				"SELECT " + getColumnsStringForSelect(childPersistentEntity) + " FROM "
-						+ childPersistentEntity.tableName() + " WHERE ");
-		StringJoiner sj = new StringJoiner(" and ");
+			SpannerPersistentEntity<T> childPersistentEntity, SpannerCustomConverter writeConverter,
+			SpannerMappingContext mappingContext) {
+		return getChildrenRowsQuery(KeySet.singleKey(parentKey), childPersistentEntity, writeConverter, mappingContext);
+	}
+
+	public static <T> Statement getChildrenRowsQuery(KeySet keySet,
+			SpannerPersistentEntity<T> childPersistentEntity, SpannerCustomConverter writeConverter,
+			SpannerMappingContext mappingContext) {
+		StringJoiner orJoiner = new StringJoiner(" OR ");
 		List<String> tags = new ArrayList<>();
 		List keyParts = new ArrayList();
 		int tagNum = 0;
 		List<SpannerPersistentProperty> childKeyProperties = childPersistentEntity
 				.getFlattenedPrimaryKeyProperties();
-		Iterator parentKeyParts = parentKey.getParts().iterator();
-		while (parentKeyParts.hasNext()) {
-			SpannerPersistentProperty keyProp = childKeyProperties.get(tagNum);
-			String tagName = "tag" + tagNum;
-			sj.add(keyProp.getColumnName() + " = @" + tagName);
-			tags.add(tagName);
-			keyParts.add(parentKeyParts.next());
-			tagNum++;
+
+		for (Key key : keySet.getKeys()) {
+			StringJoiner andJoiner = new StringJoiner(" AND ", "(", ")");
+			Iterator parentKeyParts = key.getParts().iterator();
+			while (parentKeyParts.hasNext()) {
+				SpannerPersistentProperty keyProp = childKeyProperties.get(tagNum % childKeyProperties.size());
```

Comment:
I don't think I understand what happens if the number of parent key parts exceeds the number of child key parts. Is that even supposed to happen?

Reply:
No, it could never happen

Reply:
Thanks! Then the modulo operation can be removed.

Reply:
It is there for a reason. When you have multiple keys in the keySet, you will get more key parts, because you have multiple parents. So wee need to repeat children parts for each parent.

Reply:
Would moving `int tagNum = 0` to within the `for` loop scope work? Then it would start at 0 for every parent key.

Reply:
That's not going to work, because tagNums have to be distinct. They are used for parameters binding.